File: lib/services/lyric_romanizer/engines/tamil_engine.dart
/// Tamil to Latin romanization engine.
///
/// Character-mapping romanizer for Tamil script (U+0B80–U+0BFF).

library;
import '../lyric_romanizer_types.dart';

class TamilRomanizer {
  const TamilRomanizer();

  static const Map<String, String> _vowels = {
    'அ': 'a', 'ஆ': 'aa', 'இ': 'i', 'ஈ': 'ee', 'உ': 'u',
    'ஊ': 'uu', 'எ': 'e', 'ஏ': 'ee', 'ஐ': 'ai', 'ஒ': 'o',
    'ஓ': 'oo', 'ஔ': 'au',
  };

  static const Map<String, String> _vowelSigns = {
    'ா': 'aa', 'ி': 'i', 'ீ': 'ee', 'ு': 'u', 'ூ': 'uu',
    'ெ': 'e', 'ே': 'ee', 'ை': 'ai', 'ொ': 'o', 'ோ': 'oo', 'ௌ': 'au',
  };

  static const Map<String, String> _consonants = {
    'க': 'ka', 'ங': 'nga', 'ச': 'sa', 'ஞ': 'nya', 'ட': 'da',
    'ண': 'na', 'த': 'tha', 'ந': 'na', 'ப': 'ba', 'ம': 'ma',
    'ய': 'ya', 'ர': 'ra', 'ல': 'la', 'வ': 'va', 'ழ': 'zha',
    'ள': 'la', 'ற': 'ra', 'ன': 'na',
  };

  Future<String> romanize(String line, RomanizeEngineContext context) async {
    final buf = StringBuffer();
    for (final ch in line.split('')) {
      if (_vowels.containsKey(ch)) {
        buf.write(_vowels[ch]);
      } else if (_consonants.containsKey(ch)) {
        buf.write(_consonants[ch]);
      } else if (_vowelSigns.containsKey(ch)) {
        buf.write(_vowelSigns[ch]);
      } else if (ch == '்') {
        // virama — no output, suppresses inherent 'a'
      } else {
        buf.write(ch);
      }
    }
    return buf.toString();
  }
}

File: android/app/src/main/kotlin/com/nomadguy/noctra/lyricromanizer/engines/TamilEngine.kt
package com.nomadguy.noctra.lyricromanizer.engines

import com.nomadguy.noctra.lyricromanizer.RomanizeEngineContext

/** Tamil to Latin romanizer. */
class TamilEngine {

    private val vowels = mapOf(
        'அ' to "a", 'ஆ' to "aa", 'இ' to "i", 'ஈ' to "ee", 'உ' to "u",
        'ஊ' to "uu", 'எ' to "e", 'ஏ' to "ee", 'ஐ' to "ai", 'ஒ' to "o",
        'ஓ' to "oo", 'ஔ' to "au"
    )

    private val vowelSigns = mapOf(
        'ா' to "aa", 'ி' to "i", 'ீ' to "ee", 'ு' to "u", 'ூ' to "uu",
        'ெ' to "e", 'ே' to "ee", 'ை' to "ai", 'ொ' to "o", 'ோ' to "oo", 'ௌ' to "au"
    )

    private val consonants = mapOf(
        'க' to "ka", 'ங' to "nga", 'ச' to "sa", 'ஞ' to "nya", 'ட' to "da",
        'ண' to "na", 'த' to "tha", 'ந' to "na", 'ப' to "ba", 'ம' to "ma",
        'ய' to "ya", 'ர' to "ra", 'ல' to "la", 'வ' to "va", 'ழ' to "zha",
        'ள' to "la", 'ற' to "ra", 'ன' to "na"
    )

    suspend fun romanize(line: String, context: RomanizeEngineContext): String {
        val buf = StringBuilder()
        for (ch in line) {
            vowels[ch]?.let { buf.append(it); return@let }
            consonants[ch]?.let { buf.append(it); return@let }
            vowelSigns[ch]?.let { buf.append(it); return@let }
            if (ch == '்') return@let  // virama — suppress inherent vowel
            buf.append(ch)
        }
        return buf.toString()
    }
}
